### ISSUE TITLE
Webpack 5 Support

### ISSUE THREAD

**Issue by Programmerino** [scalacenter/scalajs-bundler#350]
As expected, scalajs-bundler will throw `Unsupported webpack major version 5` when trying to use any `webpack` version with a major version higher than 4.

The update seems to primarily focus on better caching for better performance, deterministic builds, smaller output sizes, and better WebAssembly support.

I'm unsure about what needs to be changed here, but I assume that the [changes to compiler hooks](https://blog.logrocket.com/new-features-in-webpack-5-2559755adf5e/) could be important.

I don't really expect support to be added until after it comes out of beta and enough plugins get up to date, but progress can be tracked here.

Migration Guide: https://webpack.js.org/migrate/5/

**Comment by cquiroz** [comment]
I have the intention to add version 5 support (Maybe we should delete support for versions 3 and older) but sure, we need to wait for a public release first.

**Comment by cquiroz** [comment]
I'm taking a look at this, the format of the stats has changed, it may take a while to get right and it maybe hard to support both webpack 4 and webpack 5

**Comment by julienrf** [comment]
> Should we consider removing support for some of the older versions?

This is fine by me, but I would like to hear from other users of scalajs-bundler :)

**Comment by easel** [comment]
So long as 5 works, I'm happy to drop 4. I do have some customizations, but they are hopefully simple enough they won't be hard to port.

**Comment by dispalt** [comment]
yeah I say we track the latest with master, and drop support for others but keep a branch around in case.  It's not like people are needing fixes that often with this since it's mainly glue code.  The only exception seems to be that sjsbundler will need some changes for scala 1.3.0 to work.

**Comment by vhiairrassary** [comment]
Following this ticket I create this PR https://github.com/scalacenter/scalajs-bundler/pull/390 (still a draft). I would love to get feedbacks from you before finishing the remaining tasks.

**Comment by vhiairrassary** [comment]
https://github.com/scalacenter/scalajs-bundler/pull/390 has been merged. @cquiroz could you publish a RC please, so users can test & we can address issues?

**Comment by povder** [comment]
Thanks @vhiairrassary for your work on #390. Do you plan to release an RC with webpack 5 support @julienrf @sjrd @cquiroz ? I'm willing to test it out.

**Comment by arashi01** [comment]
@julienrf @sjrd @cquiroz Also happy to do any testing, or anything else needed to help get this released. Getting to the point where a number of Webpack plugins are no longer supporting V4 in their most recent iterations.

**Comment by evbo** [comment]
would love to test out the [latest RC](https://github.com/scalacenter/scalajs-bundler/releases/tag/v0.21.0-RC1) which contains webpack 5 support. Any instructions on how to manually build scalajs-bundler?

**Comment by vhiairrassary** [comment]
@evbo You can clone the repository, run `sbt publishLocal` and use the produced artefact in the `plugins.sbt` file from your project

**Comment by bblfish** [comment]
Would it be worth updating the documentation? For the moment it only mentions web pack 4 support.

**Comment by evbo** [comment]
so far for me fast optimization is working, but as @Jaystified mentioned there's some painless webpack config updates anyone using webpack needs to consider.

However, one unusual issue is after fast optimization, the dev server appears to start but with some errors:

> [webpack-cli] Unable to load '@webpack-cli/serve' command
[webpack-cli] TypeError: options.forEach is not a function

I'm also noticing `version in webpack` appears deprecated... Some other way to set that?

**Comment by unarist** [comment]
We've also encountered issues which will be fixed in https://github.com/scalacenter/scalajs-bundler/pull/408:

```
[error] Error parsing webpack stats output
[error] /warnings(0)/moduleName: JsonValidationError(List(error.path.missing),WrappedArray())
[error] /warnings(1)/moduleName: JsonValidationError(List(error.path.missing),WrappedArray())
[error] /warnings(2)/moduleName: JsonValidationError(List(error.path.missing),WrappedArray())
[error] Error parsing webpack stats output
[error] /warnings(0)/moduleName: JsonValidationError(List(error.path.missing),WrappedArray())
[error] /warnings(1)/moduleName: JsonValidationError(List(error.path.missing),WrappedArray())
[error] /warnings(2)/moduleName: JsonValidationError(List(error.path.missing),WrappedArray())
[error] java.lang.AssertionError: assertion failed: Webpack failed to create application bundle
```

Since our warnings are due to the [Performance Hints](https://webpack.js.org/configuration/performance/), we decided to use `performance: { hints: false }` as a workaround for now.

**Comment by silizza** [comment]
Hello.
Any updates here?

**Comment by matthughes** [comment]
I think the community is moving off this plugin and towards using Vite: https://www.scala-js.org/doc/tutorial/scalajs-vite.html.
